src/mempool.rs

```diff
@@ -20,225 +17,137 @@ pub enum MempoolMessage {
 /// received over the network are queued in the `Mempool` before being added to
 /// the `Blockchain`
 pub struct Mempool {
-    broadcast_channel_sender: Option<broadcast::Sender<SaitoMessage>>,
-    mempool_channel_sender: Option<mpsc::Sender<MempoolMessage>>,
-    blocks: Vec<Block>,
+    blocks: AHashMap<SaitoHash, Block>,
 }
 
 impl Mempool {
+    #[allow(clippy::clippy::new_without_default)]
     pub fn new() -> Self {
         Mempool {
-            broadcast_channel_sender: None,
-            mempool_channel_sender: None,
-	    blocks: vec![],
+            blocks: AHashMap::new(),
         }
     }
 
-    pub fn add_block(&mut self, block: Block) -> bool {
-
-	let hash_to_insert = block.get_hash();
-
-        for blk in &self.blocks {
-            if blk.get_hash() == hash_to_insert {
-        	return false;
-            }
-        }
-
-	self.blocks.push(block);
-        return true;
-
+    pub fn add_block(&mut self, block: Block) {
+        // adds block if it doesn't already exist,
+        // else it ignores attempt to add block to AHashMap
+        self.blocks.entry(block.get_hash()).or_insert(block);
```

Comment:
My intuition is that methods like this can cause bugs. If we're adding a block, it is better to know whether it needs to be added or not. If we are calling "add_block" for a block that has already been added, I think this is an indication that the calling code is unsure what it's doing. There may be special cases where you'd want this kind of "try" behaviour, but they are rare. Since we're not calling this function yet anywhere, it's hard to evaluate. I'd rather just insert or panic for now.

Reply:
A sanity check on not adding is presumably against:

 - someone generates a block with a hash collisions (improbable but it can happen -- tough luck, your candidate block is invalid)
 - some peculiar race condition it might be hard to spot where a duplicate block gets into the mempool queue while the first block is being processed but after it has left the mempool?
 
I agree with Clay that the design should not be passing blocks to add_block() unless we are sure we want to add them. Personally suspect the filtering should primarily happen in peer / mempool. Perhaps we should avoid the "ignores attempt" uncertaintly by explicitly sending the block to addBlocktoBlockchainFailure (or whatever the equivalent should be in Rust)?